### PR TITLE
TCP checker

### PR DESCRIPTION
**Pull request by emilingerslev**
Adding a TCP Checker for keeping an eye on open ports

### CODE REVIEW

File: internal/tcpchecker/checkport.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+		})
+		return "", err
+	}
+
+	scanner := bufio.NewScanner(conn)
+
+	output := ""
+	for scanner.Scan() {
+		output += fmt.Sprintf("%s\n", scanner.Text())
+	}
```
Comment (Crevil) [inline internal/tcpchecker/checkport.go]: This might be possible to simplify with a `strings.Builder` instead. The formatting and `.Text()` casting seems a but unusual to me, but dunno.

File: internal/tcpchecker/checkport.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	if err := scanner.Err(); err != nil {
+		if netError, ok := err.(net.Error); ok && netError.Timeout() {
```
Comment (Crevil) [inline internal/tcpchecker/checkport.go]: To be a bit more resilient we can use Go 1.13's `errors.As()` instead. It will unwrap the error chain and assert instead of only doing so on the first.

```golang
var netError net.Error
if errors.As(err, &netError) && netError.Timeout() {
}
```

File: internal/tcpchecker/checkport.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+			reporter.ReportPortCheck(Report{
+				Address: address,
+				Port:    port,
+				Open:    true,
+				Status:  "Open",
```
Comment (Crevil) [inline internal/tcpchecker/checkport.go]: Should this mention a timeout instead of `Open`?

File: internal/tcpchecker/logreporter.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+type LogReporter struct {
+	lastReport time.Time
+	lastOpen   bool
+	Log        interface {
+		Infof(msg string, args ...interface{})
+		Debugf(msg string, args ...interface{})
+	}
+}
+
+func (r *LogReporter) ReportPortCheck(report Report) {
```
Comment (Crevil) [inline internal/tcpchecker/logreporter.go]: Could this normalization of the state be moved down a level in the abstraction? It seems to me that it is more close to the interface one would expect to be reported than to the reporter to understand. Does that even make sense?

So along `CheckPort`s reporting of type `Report` it would have a `Transition` or `Event` or something with values 
`closed`, `opened` etc.

```
Report{
	Address: address,
	Port:    port,
	Open:    false,
	Event:   "still open",
	Status:  "Scanner error",
	Error:   err,
	Content: output,
}
```

That would reduce the complexity of this switch and let et be reused between a `LogReporter` and `Prometheus` reporter.

It could be a method on `Report` as well if that makes more sense as you do some comparison to an early `Report` from how I read it.

File: internal/tcpchecker/portchecker.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	address string
+	port    int
+}
+
+func StartPortChecking(address string, port int, reporter Reporter) PortChecker {
+
+	log.Infof("Start checking address %s:%v", address, port)
+
+	go func() {
+		for true {
```
Comment (Crevil) [inline internal/tcpchecker/portchecker.go]: Any reason to not use a `daemon.Daemon` for this loop?

Comment (emilingerslev) [reply internal/tcpchecker/portchecker.go]: That's the next step. Just wanted it working first. Will do that before moving out of draft 👍

File: internal/tcpchecker/checkport.go (hunk 3 above)
Comment (emilingerslev) [reply internal/tcpchecker/checkport.go]: The timeout is expected like with `nc -v -w1 ip:port`. Since the TCP is never closed, the timeout while reading is a success.

File: internal/tcpchecker/checkport.go (hunk 1 above)
Comment (emilingerslev) [reply internal/tcpchecker/checkport.go]: Will try that 👍

File: internal/tcpchecker/checkport.go (hunk 2 above)
Comment (emilingerslev) [reply internal/tcpchecker/checkport.go]: Ok, will have a look at that 👍

File: internal/tcpchecker/checkport.go (hunk 3 above)
Comment (Crevil) [reply internal/tcpchecker/checkport.go]: Ah, makes sense 👍

File: internal/tcpchecker/checkport.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		})
+		return "", err
+	}
+
+	scanner := bufio.NewScanner(conn)
+
+	output := strings.Builder{}
+	for scanner.Scan() {
+		fmt.Fprintf(&output, "%s\n", scanner.Text())
+	}
```
Comment (Crevil) [inline internal/tcpchecker/checkport.go]: Actually the `strings.Builder` does not really give us an advantage over the string concatenation. It just makes it harder to understand IMO. :/ Maybe we should kist stock to that?

File: internal/tcpchecker/portchecker.go [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+	"github.com/lunarway/strong-duckling/internal/daemon"
+	"github.com/prometheus/common/log"
+)
+
+type PortChecker struct {
+	address string
+	port    int
+}
+
+func StartPortChecking(address string, port int, reporter Reporter) *daemon.Daemon {
```
Comment (Crevil) [inline internal/tcpchecker/portchecker.go]: Maybe this should be renamed to something like `New*` now that it does not actually start anything?
```suggestion
func NewPortCheckerDaemon(address string, port int, reporter Reporter) *daemon.Daemon {
```

Or as the package is named `tcpchecker` we could reduce it to:
```suggestion
func NewDaemon(address string, port int, reporter Reporter) *daemon.Daemon {
```

Usage would then be:
```diff
- tcpcheker.StartPortChecking(address, port, reporter)
+ tcpchecker.NewDaemon(address, port, reporter)
```

I'm unsure if the package would ever allow a non-port checker? As is the reason to omit the `port` section of the name.

File: internal/tcpchecker/portchecker.go [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+	logger := log.With("name", "portchecker").With("address", address).With("port", port)
+	logger.Infof("Start checking address %s:%v", address, port)
+	daemon := daemon.New(daemon.Configuration{
+		Logger:   logger,
+		Interval: 1 * time.Second,
+		Tick: func() {
+			str, err := CheckPort(address, int(port), reporter)
+			logger.Debugf("Output: %s. Err %s", str, err)
```
Comment (Crevil) [inline internal/tcpchecker/portchecker.go]: Should this maybe be split into an output debug log and an error `err` log?

```suggestion
			if err != nil {
				logger.Errorf("tcpchecker on %s:%d failed: %+v", address, port, err)
			}
			logger.Debugf("tcpchecker output for  %s:%d: %s", address, port, str)
```

File: internal/tcpchecker/portchecker.go [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+func StartPortChecking(address string, port int, reporter Reporter) *daemon.Daemon {
+	logger := log.With("name", "portchecker").With("address", address).With("port", port)
+	logger.Infof("Start checking address %s:%v", address, port)
+	daemon := daemon.New(daemon.Configuration{
+		Logger:   logger,
+		Interval: 1 * time.Second,
+		Tick: func() {
+			str, err := CheckPort(address, int(port), reporter)
```
Comment (Crevil) [inline internal/tcpchecker/portchecker.go]: Why is the `int` cast necessary?

File: internal/tcpchecker/portchecker.go (hunk 7 above)
Comment (emilingerslev) [reply internal/tcpchecker/portchecker.go]: 👍

File: internal/metrics/metrics.go [hunk 10]
```diff
@@ -14,3 +16,103 @@ func Register(serveMux *http.ServeMux) {
...
+	strongDucklingVersion = "version"
+)
+
+type PrometheusReporter struct {
+	registry   prometheus.Registerer
+	version    *prometheus.GaugeVec
+	TcpChecker *TcpChecker
+}
+
+type TcpChecker struct {
```
Comment (Crevil) [inline internal/metrics/metrics.go]: Could we make this non-exported? It is an implementation detail of the package and is not useful outside in it self?

File: internal/metrics/metrics.go [hunk 11]
```diff
@@ -14,3 +16,103 @@ func Register(serveMux *http.ServeMux) {
...
+			Namespace: namespace,
+			Name:      "info",
+			Help:      "Version info of strong_duckling",
+		}, []string{strongswanVersion, strongDucklingVersion}),
+		TcpChecker: &TcpChecker{
+			open: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+				Namespace: namespace,
+				Subsystem: subSystemTcpChecker,
+				Name:      "open_info",
+				Help:      "Is TCP open is 1 otherwise 0",
```
Comment (Crevil) [inline internal/metrics/metrics.go]: Usually prom help texts are more like a sentence.

```suggestion
				Help:      "Indicates if TCP address:port is open with 1 and closed with 0",
```

I think by convention it is expected that 1 is open and 0 is not, but that is up to you to decide :)

File: internal/metrics/metrics.go [hunk 12]
```diff
@@ -14,3 +16,103 @@ func Register(serveMux *http.ServeMux) {
...
+	// case !report.Open && !r.lastOpen:
+	// 	// Port still closed
+	// 	if time.Now().Sub(r.lastReport) > 5*time.Minute {
+	// 		r.lastReport = time.Now()
+	// 		r.lastOpen = report.Open
+	// 		r.Log.Infof("Port %s:%v is still %v", report.Address, report.Port, "closed")
+	// 	}
+	// default:
+	// 	panic("This should never happen in LogReporter.ReportPortCheck")
+	// }
```
Comment (Crevil) [inline internal/metrics/metrics.go]: Could this be removed again?

File: internal/metrics/metrics_test.go [hunk 13]
```diff
@@ -0,0 +1,5 @@
+package metrics
+
+import "github.com/lunarway/strong-duckling/internal/tcpchecker"
+
+var _ tcpchecker.Reporter = &PrometheusReporter{}
```
Comment (Crevil) [inline internal/metrics/metrics_test.go]: Could we add some tests for the logic in `ReportPortCheck`? Multiple nested `if` statements are hard to maintain without along with verifying edge cases.

File: internal/tcpchecker/tcpchecker.go [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+	reporter.ReportPortCheck(Report{
+		Name:    name,
+		Address: address,
+		Port:    port,
+		Open:    true,
+		Status:  "Open",
+		Error:   nil,
+		Content: output.String(),
+	})
+	return
```
Comment (Crevil) [inline internal/tcpchecker/tcpchecker.go]: This `return` is redundant, no?

File: main.go [hunk 15]
```diff
@@ -22,18 +26,26 @@ var (
 
 func main() {
 	flags := kingpin.New("strong-duckling", "A small sidekick to strongswan VPN")
-	listenAddress := flags.Flag("listen", "Address on which to expose metrics.").Default(":9100").String()
+	listenAddress := flags.Flag("listen", "Address on which to expose metrics.").String()
 	whoopingAddress := flags.Flag("whooping", "Address on which to start whooping.").String()
+	tcpCheckerAddresses := flags.Flag("tcp-checker", "TCP address to check. Supports <address>:<port> or <name>:<address>:<port>").Strings()
```
Comment (Crevil) [inline main.go]: Would `check-tcp` be more readable on usage?

````
- strong-duckling --tcp-checker 1.2.3.4:5
+ strong-duckling --check-tcp 1.2.3.4:5

File: main.go [hunk 16]
```diff
@@ -55,12 +67,63 @@ func main() {
...
+		if len(values) == 3 {
+			name = values[0]
+			address = values[1]
+			portStr = values[2]
+		} else if len(values) == 2 {
+			address = values[0]
+			portStr = values[1]
+			name = fmt.Sprintf("%s:%s", address, portStr)
+		} else {
+			log.Errorf("Could not understand tcp-checker %s", tcpCheckerAddress)
```
Comment (Crevil) [inline main.go]:
```suggestion
			log.Errorf("Could not parse tcp-checker argument %s", tcpCheckerAddress)
```

File: strongswan/prefix-log [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+
+# setup fd-3 to point to the original stdout
+exec 3>&1
+# setup fd-4 to point to the original stderr
+exec 4>&2
+
+# get the prefix from SUPERVISOR_PROCESS_NAME environement variable
+printf -v PREFIX "%-18.18s" ${SUPERVISOR_PROCESS_NAME}
+
+# reassign stdout and stderr to a preprocessed and redirected to the original stdout/stderr (3 and 4) we have create eralier
```
Comment (Crevil) [inline strongswan/prefix-log]:
```suggestion
# reassign stdout and stderr to a preprocessed and redirected to the original stdout/stderr (3 and 4) we have create earlier
```

File: strongswan/prefix-log [hunk 18]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+
+# setup fd-3 to point to the original stdout
```
Comment (Crevil) [inline strongswan/prefix-log]: Would be nice with a short description of what to expect from this file. Any specific reason for writing this in perl instead of bash as we usually use for scripting?

File: strongswan/server.js [hunk 19]
```diff
@@ -6,4 +6,21 @@ const requestListener = function (req, res) {
...
+  const requestListener = function (req, res) {
+    res.writeHead(200);
+    res.end('Temporary hello from ' + process.env.VPN_LOCAL_PEER + "\n");
+  }
+  const server = http.createServer(requestListener);
+  const startedServer = server.listen(8090);
+  console.log("Started server")
+  await new Promise(a => setTimeout(a, time))
+  startedServer.close()
+  console.log("Stoped server")
```
Comment (Crevil) [inline strongswan/server.js]:
```suggestion
  console.log("Stopped server")
```

File: internal/tcpchecker/logreporter.go [hunk 20]
```diff
@@ -0,0 +1,46 @@
+package tcpchecker
+
+import (
+	"time"
+
+	"github.com/prometheus/common/log"
+)
+
+type LogReporter struct {
```
Comment (Crevil) [inline internal/tcpchecker/logreporter.go]: Maybe we should move the reporter implementations into a `reporter` package instead? We can do that in another PR, just wanted to note it.

We now have a `CompositeReporter`, `LogReporter` and `PrometheusReporter`. That smells like a `reporter` package with `reporter.Composite`, `reporter.Log` and `reporter.Promtheus` that can then be configured as one expects.

File: main.go [hunk 21]
```diff
@@ -55,12 +67,63 @@ func main() {
...
+			Logger:   logger,
+			Interval: 1 * time.Second,
+			Tick: func() {
+				tcpchecker.Check(name, address, int(port), tcpchecker.CompositeReporter{
+					Reporters: []tcpchecker.Reporter{
+						&tcpchecker.LogReporter{
+							Logger: logger,
+						},
+						reporter.TcpChecker,
+					},
```
Comment (Crevil) [inline main.go]: Isn't `metrics.PrometheusReporter` missing here?

File: internal/tcpchecker/compositereporter.go [hunk 22]
```diff
@@ -0,0 +1,11 @@
+package tcpchecker
+
+type CompositeReporter struct {
+	Reporters []Reporter
+}
```
Comment (Crevil) [inline internal/tcpchecker/compositereporter.go]: A small not on this API. If we use `type CompositeReporter []Reporter` usage of the reporter is a bit simpler IMO.

```diff
- reporters := tcpchecker.CompositeReporter{
- 	Reporters: []tcpchecker.Reporter{
- 		&prometheusReporter,
- 	},
- }
+ reporters := tcpchecker.CompositeReporter(&prometheusReporter)
```

File: internal/metrics/metrics.go (hunk 12 above)
Comment (emilingerslev) [reply internal/metrics/metrics.go]: Haha, hell yes. (Read you own PR first 😅)

File: strongswan/prefix-log (hunk 18 above)
Comment (emilingerslev) [reply strongswan/prefix-log]: Only reason is it is stolen from somebody. Will write a section with a reference 👍

File: internal/tcpchecker/compositereporter.go (hunk 22 above)
Comment (emilingerslev) [reply internal/tcpchecker/compositereporter.go]: Yeah let's do that with `...args` 👍

File: internal/metrics/metrics.go (hunk 10 above)
Comment (emilingerslev) [reply internal/metrics/metrics.go]: I guess so. Just tried to hide it behind a method that return the `tcpchecker.Reporter`, will push in a minute.

File: internal/tcpchecker/tcpchecker.go (hunk 14 above)
Comment (emilingerslev) [reply internal/tcpchecker/tcpchecker.go]: yes :)

File: strongswan/prefix-log (hunk 18 above)
Comment (emilingerslev) [reply strongswan/prefix-log]: Description added 👍

File: main.go (hunk 21 above)
Comment (emilingerslev) [reply main.go]: That's the `reporter.TcpChecker`, right? Maybe the variable should be `prometheusReporter` instead.

Comment (Crevil) [reply main.go]: Ah, might be. Renaming would make sense 👍

**Comment by emilingerslev** [comment]
Due to brainfarting and push'ing most this PR to master, could we maybe just merge this?
I'll do separate PR's on tests
